### ISSUE TITLE
Iterator garbage collected without being closed

### ISSUE THREAD

**Issue by papachoco** [Pylons/webtest#168]
While running nose2 tests in our application we are getting the following error

Exception AssertionError: AssertionError(u'Iterator garbage collected without being closed',) in <bound method IteratorWrapper.__del__ of <webtest.lint.IteratorWrapper object at 0x110f15690>> ignored

Though the error is ignored for a particular test; running a series layered tests this error may affect the outcome of the test run

Carlos

**Comment by papachoco** [comment]
This occurs in WebTest-2.0.24

**Comment by hannosch** [comment]
I'm also running into this with 204 responses. I'm not sure if this is related to a change in WebTest or a change in WebOb.

But the webtest tests also show this error on all Python versions, from travis-ci (https://travis-ci.org/Pylons/webtest/builds/191495513):
```
test_204 (tests.test_debugapp.TestTesting) ...
/home/travis/build/Pylons/webtest/webtest/lint.py:479:
WSGIWarning: Content-Type header found in a 204 response, which not return content.

Exception AssertionError: AssertionError(u'Iterator garbage collected without being closed',) in
<bound method IteratorWrapper.__del__ of <webtest.lint.IteratorWrapper object at 0x7ff0ec1931d0>> ignored
```

204 responses have no body, so it might make sense that no close method is being called here and the lint check might be too aggressive. It seems webob sets the `response.body` to ``[b'']`` for these and tries to delete the content-type header. But how exactly it does this has changed in 1.7.0 and there's a changelog entry for it in the [backwards incompatibility](https://github.com/Pylons/webob/blob/1.7-branch/CHANGES.txt) section.

**Comment by bertjwregeer** [comment]
This is something that is going to be fixed in WebOb: https://github.com/Pylons/webob/issues/305

The issue is that when `app_iter` is provided it is not being set on the `Response`. This will get fixed so that if `app_iter` is set, it will be set in the `Response`. If that `app_iter` contains data when it shouldn't (for example because it is a 204 response) then the user bought it because `app_iter` is considered advanced usage of the API.

I apologise for the breakage. Will get a new WebOb release out shortly to fix the issue.

**Comment by bertjwregeer** [comment]
So looking at https://github.com/Pylons/webtest/blob/master/webtest/debugapp.py#L59 it looks like it sets the body to `b''` and then creates some headers: https://github.com/Pylons/webtest/blob/master/webtest/debugapp.py#L62 and assigns those headers here: https://github.com/Pylons/webtest/blob/master/webtest/debugapp.py#L77

WebOb won't try and stop you from shooting yourself in the foot, and updating headers with ones that are invalid for the status or having a body on a non-body Response are not things WebOb tries to stop the user from doing.

**Comment by gawel** [comment]
I don't think the header warning is the issue. We have this one for a very long time. Because webtest warn you when you shoot yourself in the foot :)

So if you're saying that you'll fix the body problem, we'll just wait and see.

**Comment by bertjwregeer** [comment]
@gawel Latest build: https://travis-ci.org/Pylons/webtest/jobs/191495514#L389 no more

```Exception AssertionError: AssertionError(u'Iterator garbage collected without being closed',) in
<bound method IteratorWrapper.__del__ of <webtest.lint.IteratorWrapper object at 0x7ff0ec1931d0>> ignored```

**Comment by bertjwregeer** [comment]
New release of WebOb is out: https://pypi.python.org/pypi/WebOb/1.7.1

**Comment by gawel** [comment]
Great. Thanks!

**Comment by hannosch** [comment]
Confirmed fixed with WebOb 1.7.1. Thanks for the quick fixes everyone!
